git-town 4.0.0

ilovezfs opened the pull request:
depends on go
no longer bottle unneeded
drop support for Yosemite and below
don't install outdated man files

ilovezfs commented:
CC @charlierudolph

ilovezfs commented:
I guess we also need to drop support for < Yosemite or depend on brew git due to https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/1309/version=yosemite/testReport/junit/brew-test-bot/yosemite/test_git_town/

ilovezfs commented:
Yosemite incompatibility reported upstream https://github.com/Originate/git-town/issues/902.

charlierudolph commented:
Thanks @ilovezfs 

We had hoped to not need to build on the users machine. We are attaching binaries to [releases](https://github.com/Originate/git-town/releases/tag/v4.0.0). Is there any way we can use those?

cc @kevgo

ilovezfs commented:
@charlierudolph that's why I've removed `bottle :unneeded`. That means we will now ship the binaries our CI builds.

charlierudolph commented:
Do we need the go dependency then? I would think that could be removed since we are just installing a binary.

ilovezfs commented:
@charlierudolph That's why it's marked with `=> :build`

charlierudolph commented:
What I'm meaning to ask is if we could remove the call to "go install" which creates the binary and instead just use one that we have on our release page.

> That means we will now ship the binaries our CI builds....That's why it's marked with => :build

Out of curiosity how does the CI build work? It appears it would need to run a portion of the install function and save a couple steps to actually run on the users machine.

ilovezfs commented:
>What I'm meaning to ask is if we could remove the call to "go install" which creates the binary and instead just use one that we have on our release page.

No, that would be contrary to Homebrew policy, which is to build our own binaries for open-source CLI tools.

>Out of curiosity how does the CI build work? It appears it would need to run a portion of the install function and save a couple steps to actually run on the users machine.

That works because I set `ENV["GOBIN"] = bin`. The bash equivalent would be `export GOBIN=/usr/local/Cellar/git-town/4.0.0/bin`.

ilovezfs commented:
Alternatively, we could do
```
diff --git a/Formula/git-town.rb b/Formula/git-town.rb
index 55c713e..7484a52 100644
--- a/Formula/git-town.rb
+++ b/Formula/git-town.rb
@@ -9,11 +9,9 @@ class GitTown < Formula
 
   def install
     ENV["GOPATH"] = buildpath
-    ENV["GOBIN"] = bin
     (buildpath/"src/github.com/Originate").mkpath
     ln_sf buildpath, buildpath/"src/github.com/Originate/git-town"
-    system "go", "install", "-v"
-    bin.install_symlink "git-town-#{version}" => "git-town"
+    system "go", "build", "-o", bin/"git-town"
     man1.install Dir["man/man1/*"]
   end
 ```

I don't know whether you care about having the versioned binary `git-town-4.0.0` or not.

charlierudolph commented:
The versioned binary is good with me. Can you please also remove the linking of the man files?Those are now outdated and should be removed.

I'd like to wait for @kevgo's thoughts before we drop yosemite. Is there anyway to specify a minimum git version? I'm guessing something that refuses to install unless git has a specific version or higher.

ilovezfs commented:
>Is there anyway to specify a minimum git version? I'm guessing something that refuses to install unless git has a specific version or higher.

Nope. We could depend on brew's git for <= Yosemite, but that's a pretty heavy dependency to tack on if it's at all possible to fix this upstream instead.

charlierudolph commented:
@ilovezfs can we add git >= 2.6.0 as a dependency? We have always had `git` as a dependency and I think specifying the version we depend on is fine. On the `git-town` side we can also add a check that the git version is that or higher as well in case we get installed via other means.

Also I like your [alternative](https://github.com/Homebrew/homebrew-core/pull/13537#issuecomment-301345700) and don't care about having a versioned binary.

ilovezfs commented:
@charlierudolph we can't depend on a specific git version, but we can do this:
```
  depends_on "git" if MacOS.version <= :yosemite
```

ilovezfs commented:
Actually, I lied. It looks like we have an audit specifically preventing that:
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dev-cmd/audit.rb#L469-L470

ilovezfs commented:
@charlierudolph I'm not thrilled with the idea, but something like this is also possible:
```
diff --git a/Formula/git-town.rb b/Formula/git-town.rb
index cb594b35bb..04ef13dcd1 100644
--- a/Formula/git-town.rb
+++ b/Formula/git-town.rb
@@ -1,3 +1,27 @@
+class Git26Requirement < Requirement
+  fatal true
+  default_formula "git"
+
+  satisfy do
+    git = which("git")
+    next unless git
+    git_version = Utils.popen_read("git --version")
+    next unless $?.exitstatus.zero?
+    git_version = git_version.lines.first.chomp.split(" ")[2].split(".")
+    next unless git_version[0].to_i >= 2
+    next unless git_version[1].to_i >= 11
+    git
+  end
+
+  def message; <<-EOS.undent
+    Git 2.6.0+ is required to install.
+
+    You can install this with:
+      brew install git
+    EOS
+  end
+end
+
 class GitTown < Formula
   desc "High-level command-line interface for Git"
   homepage "http://www.git-town.com"
@@ -5,7 +29,7 @@ class GitTown < Formula
   sha256 "4145a77ee7fe1f4579d04dfcc0a4658a21db4e07e80d7dd79328162f12797193"
 
   depends_on "go" => :build
-  depends_on :macos => :el_capitan
+  depends_on Git26Requirement
 
   def install
     ENV["GOPATH"] = buildpath
```

ilovezfs commented:
@charlierudolph only 9 of 388 unique installs in the last 90 days were on <= Yosemite, so I think we should just drop support.

ilovezfs commented:
@charlierudolph @kevgo I'm pulling this without Yosemite support, but happy to add it back if compatibility with < git 2.6.0 is restored.
